securedrop_client/gui/state.py

```diff
@@ -0,0 +1,44 @@
+import textwrap
+from typing import Dict, List, NewType, Optional
+from uuid import UUID
+
+from PyQt5.QtCore import QObject
+
+ConversationId = NewType("ConversationId", UUID)
+FileId = NewType("FileId", UUID)
+
+
+class State(QObject):
+    def __init__(self) -> None:
+        super().__init__()
+
+        self._selected_conversation: Optional[ConversationId] = None
+        self._conversations: Dict[ConversationId, List[FileId]] = {}
+
+    def selected_conversation(self) -> Optional[ConversationId]:
+        return self._selected_conversation
+
+    def conversation_files(self, id: ConversationId) -> List[FileId]:
+        return self._conversations.get(id, [])
+
+    def __repr__(self) -> str:
+        return textwrap.dedent(
+            f"""
+                selected_conversation: {self._selected_conversation}
+                conversation_files: {self._conversations}
+            """
+        )
+
+
+class Writer:
```

Comment:
methods here probably should be moved into the State class and you can use setters if ya like

Reply:
This split was intentional, let me write that down :slightly_smiling_face: 

Currently, the `controller` is pulled in many UI widgets for two reasons:
1. connecting to signals emitted by the controller (I think of it as a "read" operation)
2. calling methods on the controller (I think of it as a "write"/"command" operation)

Ideally, the "write" operations would only happen in _actions_ (`QAction`), which would form the interface between the UI and the controller (which is not part of the `gui` package). The other widgets could still subscribe to the changes of state by consuming the corresponding controller signals ("read" operations).

By separating the read and write interfaces, it would become considerably more difficult to mix those accidentally, since _actions_ and _the other widgets_ would be using references to different objects (currently called `gui_state_writer` and `gui_state` in `securedrop_client/app.py`).

Reply:
In other words, the idea is to have a class (`state.State`) with a collection of signals and no public methods, and another class (`state.Writer`) with a collection of public methods. The actions would use those methods to issue commands to the controller.

Reply:
@creviera Just taking note there that we may revisit the split between reader and writer. :spiral_notepad:

Reply:
Update here: I'm merging back reader and writer. It will be easy enough to make the change later and I feel like the extra diff length is not fully justified right now.